sky/packages/sky/lib/src/fn3/popup_menu.dart

```diff
@@ -0,0 +1,206 @@
+
+import 'package:sky/animation.dart';
+import 'package:sky/painting.dart';
+import 'package:sky/material.dart';
+import 'package:sky/src/fn3/basic.dart';
+import 'package:sky/src/fn3/focus.dart';
+import 'package:sky/src/fn3/framework.dart';
+import 'package:sky/src/fn3/gesture_detector.dart';
+import 'package:sky/src/fn3/navigator.dart';
+import 'package:sky/src/fn3/popup_menu_item.dart';
+import 'package:sky/src/fn3/scrollable.dart';
+import 'package:sky/src/fn3/transitions.dart';
+
+const Duration _kMenuDuration = const Duration(milliseconds: 300);
+const double _kMenuCloseIntervalEnd = 2.0 / 3.0;
+const double _kMenuWidthStep = 56.0;
+const double _kMenuMargin = 16.0; // 24.0 on tablet
+const double _kMenuMinWidth = 2.0 * _kMenuWidthStep;
+const double _kMenuMaxWidth = 5.0 * _kMenuWidthStep;
+const double _kMenuHorizontalPadding = 16.0;
+const double _kMenuVerticalPadding = 8.0;
+
+class PopupMenu extends StatefulComponent {
+  PopupMenu({
+    Key key,
+    this.items,
+    this.level: 4,
+    this.navigator,
+    this.performance
+  }) : super(key: key) {
+    assert(items != null);
+    assert(performance != null);
+  }
+
+  final List<PopupMenuItem> items;
+  final int level;
+  final NavigatorState navigator;
+  final WatchableAnimationPerformance performance;
+
+  PopupMenuState createState() => new PopupMenuState(this);
+}
+
+class PopupMenuState extends ComponentState<PopupMenu> {
+  PopupMenuState(PopupMenu config) : super(config) {
+    _updateBoxPainter();
+    config.performance.addListener(_performanceChanged);
+  }
+
+  BoxPainter _painter;
+
+  void _updateBoxPainter() {
+    _painter = new BoxPainter(
+      new BoxDecoration(
+        backgroundColor: Colors.grey[50],
+        borderRadius: 2.0,
+        boxShadow: shadows[config.level]
+      )
+    );
+  }
+
+  void didUpdateConfig(PopupMenu oldConfig) {
+    if (config.level != config.level)
+      _updateBoxPainter();
+    if (config.performance != oldConfig.performance) {
+      oldConfig.performance.removeListener(_performanceChanged);
+      config.performance.addListener(_performanceChanged);
+    }
+  }
+
+  void dispose() {
+    config.performance.removeListener(_performanceChanged);
+  }
+
+  void _performanceChanged() {
+    setState(() {
+      // the performance changed, and our state is tied up with the performance
+    });
+  }
+
+  Widget build(BuildContext context) {
+    double unit = 1.0 / (config.items.length + 1.5); // 1.0 for the width and 0.5 for the last item's fade.
+    List<Widget> children = [];
+    for (int i = 0; i < config.items.length; ++i) {
+      double start = (i + 1) * unit;
+      double end = (start + 1.5 * unit).clamp(0.0, 1.0);
+      children.add(new FadeTransition(
+        performance: config.performance,
+        opacity: new AnimatedValue<double>(0.0, end: 1.0, interval: new Interval(start, end)),
+        child: new GestureDetector(
+          onTap: () { config.navigator.pop(config.items[i].value); },
```

Comment:
I think eventually we'll want PopupMenu to accept two kinds of children, boring PopupMenuItems that have a value and do this, and some sort of StatefulPopupMenuItem that can do more complicated things like have submenus, colour palettes, etc, and does its own navigator popping (probably by finding the PopupMenu itself). But for now this is fine.